feat(server): add http api

ds-skoropad opened the pull request:
Приветствую!

Изначально загорелся желанием реализовать через некий билдер , что после вызвало у меня некие сомнения.
В целом код рабочий. И если выражение "Я художник, я так вижу" меня никак не умоляет, значит будем каким-то образом переписывать :)

Как и всегда готов к любым замечаниям.

ds-skoropad committed:
feat(server): add http api

Sprint 9 Review 1

ds-skoropad committed:
style(main): fix the modifier

Sprint 9 Review 1

Ivanov-s-tmn commented on java-kanban.iml:
файл с расширением .iml стоит добавить в gitignore, это локальный файл настроек, который генерируется ide и не нужен в репозитории

Ivanov-s-tmn commented on src/ru/yandex/practicum/kanban/http/HttpTaskServer.java:
```diff
@@ -0,0 +1,124 @@
+                .create();
+    }
+
+    private static class StatusTaskAdapter extends TypeAdapter<StatusTask> {
+
+        @Override
+        public void write(final JsonWriter jsonWriter, final StatusTask statusTask) throws IOException {
+            jsonWriter.value(statusTask.name());
+        }
+
+        @Override
+        public StatusTask read(final JsonReader jsonReader) throws IOException {
+            String statusTask = jsonReader.nextString();
+            if (TaskUtils.stringInArrEnum(StatusTask.values(), statusTask)) {
+                return StatusTask.valueOf(statusTask);
+            } else {
+                throw new HttpServerGsonException("Json field parsing error: status=" + statusTask);
+            }
+        }
+    }
```
А зачем адаптер для статусов? Вроде с ними не должно быть проблем, обычный енам и он легко преобразуется в строку

ds-skoropad replied:
Да все верно, с ним нет проблем. Добавлял когда тестировал возможность передачи json с неверным статусом, например "status=NEWand345", по ТЗ такое не предусматривается. Должен был удалить этот адаптер после :)
Убрал из кода.

Ivanov-s-tmn commented on src/ru/yandex/practicum/kanban/http/HttpTaskServer.java:
```diff
@@ -0,0 +1,124 @@
+    private static class LocalDateTimeAdapter extends TypeAdapter<LocalDateTime> {
+
+        @Override
+        public void write(final JsonWriter jsonWriter, final LocalDateTime localDateTime) throws IOException {
+            if (localDateTime == null) {
+                jsonWriter.value("null");
+            } else {
+                jsonWriter.value(localDateTime.format(ISO_LOCAL_DATE_TIME));
+            }
+        }
+
+        @Override
+        public LocalDateTime read(final JsonReader jsonReader) throws IOException {
+            try {
+                String localDateTime = jsonReader.nextString();
+                return (localDateTime.equals("null") || localDateTime.isEmpty())
+                        ? null : LocalDateTime.parse(localDateTime, ISO_LOCAL_DATE_TIME);
+            } catch (DateTimeParseException e) {
+                throw new HttpServerGsonException(e.getMessage());
+            }
+        }
+    }
+
+    private static class DurationAdapter extends TypeAdapter<Duration> {
+
+        @Override
+        public void write(final JsonWriter jsonWriter, final Duration duration) throws IOException {
+            jsonWriter.value(duration.getSeconds());
+        }
+
+        @Override
+        public Duration read(final JsonReader jsonReader) throws IOException {
+            String duration = jsonReader.nextString();
+            if (TaskUtils.isNumber(duration)) {
+                return Duration.ofSeconds(Long.parseLong(duration));
+            } else {
+                throw new HttpServerGsonException("Json field parsing error: duration=" + duration);
+            }
+        }
+    }
```
Лучше вынести в отдельные классы, можно выделить отдельный пакет, например adapter)

Ivanov-s-tmn commented on src/ru/yandex/practicum/kanban/http/handler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,84 @@
+import java.util.List;
+import java.util.Optional;
+
+public class TaskHttpHandler extends BaseHttpHandler implements HttpHandler {
```
Не лучше в `BaseHttpHandler` имплементировать интерфейс `HttpHandler`, сделать класс абстрактным, а в конкретных хендлерах уже писать реализацию? 
Поля gson и manager можно тоже вынести в абстрактный класс

Ivanov-s-tmn commented on test/ru/yandex/practicum/kanban/http/HttpTaskServerEpicTaskTest.java:
```diff
@@ -0,0 +1,222 @@
+    static class SubtaskListTypeToken extends TypeToken<List<SubTask>> {
+    }
+
+    @BeforeEach
+    void setUp() {
+        manager.clearTaskGroup();
+        manager.clearSubGroup();
+        manager.clearEpicGroup();
+        taskServer.start();
+    }
+
+    @AfterEach
+    void tearDown() {
+        taskServer.stop();
+    }
```
Можно тоже сделать какой-то базовый класс для тестов и вынести туда основную логику и хуки

ds-skoropad replied:
Попытка написать общие тесты для всех типов задач привела к нарастанию кода при определении запуска соответствующего метода. Плюс ошибки gson, например вот так не работает:

```
public abstract class HttpTaskServerTest<T extends Task>...
public static class TaskListTypeToken extends TypeToken<List<T>>
```
gson формирует список с поломанным обьектом.

Это все удалось обойти, но по итогу при описании логики тестирования пересечений и вызова соответствующих методов код становился трудно читаемым или громозким.

В результате для простоты перенес в абстракцию только общие методы.

ds-skoropad committed:
chore(gitignore): exclude .iml

Sprint 9 Review 2

ds-skoropad committed:
refactor(http): remove StatusTaskAdapter

The conversion control of the "StatusTask" field is not required.
Sprint 9 Review 2

ds-skoropad committed:
refactor(http): transfer adapters to a package

Sprint 9 Review 2

ds-skoropad committed:
refactor(http): make the base handler abstract

Sprint 9 Review 2

ds-skoropad committed:
test(http): add an abstract class for tests

Sprint 9 Review 2

ds-skoropad committed:
style(http): make a separation

Sprint 9 Review 2
